[IV-24-25] Objetivo 2

FerniCuesta opened the pull request:
**Sobre la estructura del repositorio**
* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

**Sobre el planteamiento**
* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

**Sobre el análisis del problema**
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor, una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo objeto valor que no dependa de ningún otro (y que sea la base de muchos otros)?.

**Sobre el código**
* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

FerniCuesta committed:
closes Carlosmapego8#5 Establece Go como lenguaje para el proyecto

FerniCuesta committed:
closes Carlosmapego8#6 Crea paquete Turno

Define un Turno en base a los requerimientos del problema

FerniCuesta committed:
close Carlosmapego8#11 Crea el paquete DiasLibres

Los días libres pueden ser por asuntos propios, por vacaciones o por baja

FerniCuesta committed:
close Carlosmapego8#8 Crea el paquete Contrato

Un contrato está formado por un tipo de contrato y un número de horas

FerniCuesta committed:
close Carlosmapego8#6 Añade el día al Turno

Añade el campo Día de tipo Time para poder identificar el día del turno

FerniCuesta committed:
close Carlosmapego8#7 Crea el paquete Auxiliar

Define de forma inequívoca a un auxiliar

FerniCuesta committed:
close Carlosmapego8#10 Añade el turno preferido a un auxiliar

Permite poder asignar un turno de preferencia a un auxiliar

FerniCuesta committed:
close Carlosmapego8#9 Crea el paquete GestorTurnos

GestorTurnos contiene los auxiliares disponibles para poder realizar el reparto

FerniCuesta committed:
Crea el módulo de la aplicación

FerniCuesta committed:
Establece la entidad en iv.yaml

Carlosmapego8 commented on internal/turno/turno.go:
```diff
@@ -0,0 +1,20 @@
+
+type Turno struct {
+	Dia time.Time
+	HoraInicio string
```
El tipo time ya almacena de por sí la fecha y la hora de un instante. ¿Qué aporta almacenar el día en este tipo y las horas de inicio y fin como un string?

FerniCuesta replied:
Resuelto en [c4187e1](https://github.com/FerniCuesta/GoMountain/commit/c4187e18084f888099f3f1b852fc31a89eb7d8f0)

Carlosmapego8 commented on internal/auxiliar/auxiliar.go:
```diff
@@ -0,0 +1,17 @@
+	Contrato contrato.Contrato
+	Turnos []turno.Turno
+	TurnoPreferido turno.TipoTurno
+	DiasLibres []diaslibres.DiasLibres
```
A excepción de los días de baja, que sí dependen en exclusiva de la situación de cada auxiliar, los días libres vienen definidos según el tipo de contrato. ¿Deberían estar los días libres relacionados con el contrato y con el auxiliar o solo con el auxiliar?

FerniCuesta replied:
Resuelto en [e22feeb](https://github.com/FerniCuesta/GoMountain/commit/e22feebbfa786e346efb85e07a41d206f107ee19)

Carlosmapego8 commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: Go
+entidad: gestorturnos.go
```
En cuanto a la estructura de carpetas creo que una carpeta internal para todos los archivos (a excepción del de módulo) sería algo más apropiado, de esta forma quedaría todo el módulo incluido en una única carpeta que gestione todo el módulo

FerniCuesta replied:
Resuelto en [6c45672](https://github.com/Carlosmapego8/GoMountain/pull/12/commits/6c45672cb5a5688a65871bbb219130e30b8dd726)

Carlosmapego8 reviewed (changes requested):
Te dejo unos cuantos comentarios con posibles mejoras

FerniCuesta committed:
Carlosmapego8#6 Establece el inicio y fin del Turno como tipo Time

FerniCuesta committed:
Carlosmapego8#7 Cambia la disposición de días libres y días de baja

Los días libres dependen del contrado, los días de baja dependen del auxiliar

FerniCuesta committed:
https://github.com/Carlosmapego8/GoMountain/pull/12#discussion_r1869969384 Mueve el programa principal a la carpeta internal

Carlosmapego8 reviewed (changes requested):
En cuanto al código ahora sí lo veo mejor. Pero en cuanto a la estructura de carpetas, me refería a tener todos los .go dentro de la misma carpeta internal, en lugar de una carpeta por archivo como está ahora mismo. Con ese simple cambio te doy el aprobado.

FerniCuesta commented:
Viendo la documentación oficial (https://go.dev/doc/modules/layout) he decidido crear carpetas distintas dentro de internal de cara a posibles futuros test. Coméntame si prefieres mantener la distribución actual o hacer la modificación que mencionas.

Carlosmapego8 commented:
Me parece bien hacerlo como lo comentas, te dejo los cambios aprobados

Carlosmapego8 commented:
- [x] ¿Los issues que ha planteado la persona son efectivamente problemas y evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
- [x] ¿Has comprobado que efectivamente el compañero/a ha seguido su lista de
comprobación?
 - [x] ¿El código creado es una modelización válida de las historias de usuario
creadas por ti, y te sirve como base para programar la lógica de negocio?
- [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
programación que hayáis consensuado, tanto en identificadores, como
nombres de ficheros, como disposición de los diferentes ficheros en el
repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
aportan código para resolver el problema planteado en el mismo?
- [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el 
- [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
- [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
- [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

FerniCuesta commented:
Buenas tardes @jj, le solicito revisión del objetivo

JJ commented on internal/auxiliar/auxiliar.go:
```diff
@@ -0,0 +1,16 @@
+type Auxiliar struct {
+	ID int
+	Nombre string
+	Apellidos string
```
¿El nombre *y* los apellidos son necesarios en un producto mínimamente viable?

FerniCuesta replied:
Al poder distinguir un Auxiliar de otro mediante el ID y no estar mencionado en las HUs qué campos de los Auxiliares se almacenan, podemos prescindir del nombre y los apellidos en el PMV.

FerniCuesta replied:
Resuelto en [5ba92a3](https://github.com/FerniCuesta/GoMountain/commit/5ba92a3e29fb5fd1c75ee08c1653c895fe09fc7a)

JJ commented on internal/auxiliar/auxiliar.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+type Auxiliar struct {
+	ID int
```
¿Esto es un objeto valor o una entidad?

FerniCuesta replied:
Se trata de una entidad ya que se identifica de forma inequívoca por su identificador. Puede compartir con otras instancias del mismo tipo el mismo valor en sus campos, pero el ID distingue a ambas. El valor de sus campos puede variar en su ciclo de vida.

JJ commented on internal/contrato/contrato.go:
```diff
@@ -0,0 +1,20 @@
+
+type Contrato struct {
+	Tipo TipoContrato
+	Horas int
```
¿Pueden ser -3 horas?

FerniCuesta replied:
Resuelto en [cd978f5](https://github.com/FerniCuesta/GoMountain/commit/cd978f5c2fa970a669b985a3d8f9b6f249aa95ea)

JJ commented on internal/diaslibres/diaslibres.go:
```diff
@@ -0,0 +1,10 @@
+type Motivo int
+
+const (
+	AsuntosPropios Motivo = iota
```
¿Esto eventualmente se va a imprimir? ¿Si se va a imprimir, es adecuado que tenga un número?

FerniCuesta replied:
Resuelto en [ed98733](https://github.com/FerniCuesta/GoMountain/commit/ed987330e39bead71f53d479119f0523f315fa2b).
Se aplica la misma solución a los tipos TipoContrato en [ce2ecc2](https://github.com/FerniCuesta/GoMountain/commit/ce2ecc2cf2a8d7eb7a0517b8bd624cdbe2b8d33d) y Turno en [5d723ad](https://github.com/FerniCuesta/GoMountain/commit/5d723adc48497c347e9eadbc9b2ec5fa929e28fb)

JJ commented on internal/auxiliar/auxiliar.go:
¿Sigue las mejores práticas de go meter cada fichero en un `package` diferente?

FerniCuesta replied:
Resuelto en [969be85](https://github.com/Carlosmapego8/GoMountain/pull/12/commits/969be85246343a3844a8e5893adf7cf3f29a5908)

JJ commented on internal/gestorturnos.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+type GestorTurnos struct {
+	Auxiliares []auxiliar.Auxiliar
```
¿Cómo garantizas que hay una única fuente de verdad para los auxiliares?

FerniCuesta replied:
Resuelto en [33b0f8a](https://github.com/FerniCuesta/GoMountain/commit/33b0f8af76201454b2d841dafd8efa8416514b58)

JJ commented on internal/contrato/contrato.go:
```diff
@@ -0,0 +1,20 @@
+	"GoMountain/internal/diaslibres"
+)
+
+type Contrato struct {
```
Esto se ha modificado en este commit 0dffb4b
Hace referencia a #8. El commit no explica por qué se ha tomado esa decisión, pero además por ningún lado menciona el tipo de contrato.

FerniCuesta replied:
Como se menciona en https://github.com/Carlosmapego8/GoMountain/issues/8, el tipo de contrato se establece en el [README](https://github.com/Carlosmapego8/GoMountain) del repositorio.

> Los contratos son de los siguientes tipos:
    Sustitutos: contratos variables según necesidad
    Rotatorios: 40 horas semanales (con o sin noches dependiendo del caso)
    Jornada Reducida: 20 horas semanales (con o sin noches dependiendo del caso)
    Contratos extraordinarios: horas extra, guardias, ...

JJ reviewed (changes requested):
Si estáis tomando el enfoque de una clase por issue, o peor, un paquete por cada clase y por cada issue, estás tomando el enfoque equivocado. Eso no es modelar, es poner nombre a las cosas y en algunos casos nombres a cosas inventadas.
Un modelo es otra cosa diferente, y @Carlosmapego8 debería intentar ver realmente como va a usar esto en el objetivo 4 antes de aprobar todo y dejarme el trabajo a mi de ver qué conviene o no a *su* proyecto. Es lo que tengo que hacer, ya que estoy evaluando a @FerniCuesta que debe aprender una serie de cosas en este objetivo, pero aquí tenéis que poner en práctica lo que hayáis aprendido en el objetivo 1 aplicando una metodología, Domain Driven Design, que haríais bien en consultar los dos.
Por otro lado, antes de la siguiente revisión os aconsejaría que leyérais los dos los errores frecuentes.

FerniCuesta committed:
Carlosmapego8#7 Elimina campos innecesarios de la entidad Auxiliar

Elimina los campos del nombre y los apellidos al tratarse se campos no necesarios en un PMV

FerniCuesta committed:
Carlosmapego8#8 Crea un constructor para el tipo Contrato

FerniCuesta committed:
https://github.com/Carlosmapego8/GoMountain/pull/12#discussion_r1872895450 Mueve todos los tipos a un mismo paquete

Todos los tipos se definen en el paquete gestorturnos.
Se modifica también la estructura de directorios, todos los tipos se encuentran en internal sin subdirectorios.

FerniCuesta committed:
Carlosmapego8#9 Crea un patron singleton para GestorTurnos

FerniCuesta committed:
Carlosmapego8#11 Cambia el tipo Motivo de int a string

De esta forma tiene el tipo adecuado para poder imprimirse

FerniCuesta committed:
Carlosmapego8#8 Cambia el tipo TipoContrato de int a string

De esta forma tiene el tipo adecuado para poder imprimirse

FerniCuesta committed:
Carlosmapego8#6 Cambia el tipo Turno de int a string

De esta forma tiene el tipo adecuado para poder imprimirse

FerniCuesta commented:
Buenas @Carlosmapego8, 
se han realizado una serie de cambios en base a los comentarios de JJ. Te pido que los revises y me des tu feedback de ellos.

Hay dos comentarios sin resolver (https://github.com/Carlosmapego8/GoMountain/pull/12#discussion_r1872891541 y https://github.com/Carlosmapego8/GoMountain/pull/12#discussion_r1872900328) ya que no se ha realizado ninguna modificación de código a partir de ellos. Si tienes algo que aportar en ellos siéntete libre de hacerlo.

Un saludo.

Carlosmapego8 reviewed (approved):
Yo, por mi parte lo veo correcto

FerniCuesta commented:
Buenas @JJ, le solicito revisión del objetivo.

JJ commented on internal/auxiliar.go:
```diff
@@ -0,0 +1,9 @@
+package gestorturnos
+
+type Auxiliar struct {
+	ID             int
```
¿Esto es un objeto valor? ¿Por qué se caracterizan los objetos valor?

FerniCuesta replied:
Modificado en https://github.com/Carlosmapego8/GoMountain/pull/12/commits/a77779c3f076ce695ecf625b0a26183166591d79

JJ commented on internal/auxiliar.go:
```diff
@@ -0,0 +1,9 @@
+type Auxiliar struct {
+	ID             int
+	Contrato       Contrato
+	Turnos         []Turno
```
Misma pregunta que arriba.

FerniCuesta replied:
Modificado en https://github.com/Carlosmapego8/GoMountain/pull/12/commits/a77779c3f076ce695ecf625b0a26183166591d79

JJ commented on internal/auxiliar.go:
```diff
@@ -0,0 +1,9 @@
+package gestorturnos
+
+type Auxiliar struct {
```
Esto está modificado en este commit [969be85](https://github.com/Carlosmapego8/GoMountain/commit/969be85246343a3844a8e5893adf7cf3f29a5908), que no hace referencia a ningún issue.
En ese caso, es imposible saber si añaden valor al cliente, porque no se ve como se corresponden con la historia de usuario.

FerniCuesta replied:
Ya ha sido relacionado con el issue correspondiente (https://github.com/Carlosmapego8/GoMountain/issues/5)

JJ replied:
Si el código está mal desarrollado, está mal desarrollado. Hacer una ñapa a posteriori no lo arregla, sigue sin haberse desarrollado siguiendo el procedimiento adecuado.
Tampoco se deben resolver los comentarios a menos que se muestre en qué commit han sido resueltos.
Y el issue que pones es en el que se elige el lenguaje de programación. *Por eso* no se deben hacer ñapas a posteriori, sino desarrollar correctamente, que es la única forma de garantizar que se hace bien.

JJ reviewed (changes requested):
Es conveniente, por parte de los dos, que se entiendan los conceptos de domain driven design, especialmente la diferencia entre objeto valor y entidad, y la importancia de la gestión de la mutabilidad y quién debe hacerlo.
Por otro lado, indicar en *cada* commit a qué issue corresponde es imprescindible, ya que si no se hace imposible saber cómo se está añadiendo valor a un cliente. También cuando se responde a comentarios se debe añadir valor al cliente, porque es quien va a comprar la aplicación, no quien hace comentarios.

FerniCuesta committed:
Carlosmapego8#7 Define un Auxiliar como objeto valor

El tipo Auxiliar se identifica de forma inequívoca por el valor de sus atributos

FerniCuesta committed:
carlosmapego8#9 Se establece una SSOT para los Auxiliares

Se emplea el tipo de dato map para a partir de un string único obtener un Auxiliar. Esto garantiza no tener el mismo Auxiliar con distintos valores.

FerniCuesta commented:
Buenas @Carlosmapego8, 

he realizado modificaciones en base a los comentarios realizados. Te solicito una nueva revisión.

Carlosmapego8 commented on internal/gestorturnos.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+type GestorTurnos struct {
+	Auxiliares map[string]Auxiliar
```
¿No crees que es redundante copiar los Auxiliares en la estructura junto al string que los identifica? ¿Se podrían recuperar los datos de un auxiliar exclusivamente a través del string?

FerniCuesta replied:
Debido a que en Go no existe de forma nativa la estructura de datos Set y un vector no nos garantiza SSOT, creo que la mejor implementación es con un Map.

Carlosmapego8 replied:
Entiendo, creo que es una solución adecuada la que planteas

Carlosmapego8 commented on internal/auxiliar.go:
```diff
@@ -0,0 +1,11 @@
+type Auxiliar struct {
+	NombreUnico    string
+	Contrato       Contrato
+	Turnos         map[time.Time]Turno
```
Misma pregunta que arriba, pero cambiando string por esta fecha

FerniCuesta replied:
Debido al mismo motivo que arriba, creo que la mejor implementación es con un Map.

Carlosmapego8 replied:
Lo mismo que arriba

JJ replied:
También tenéis que resolver *vuestros* comentarios con un commit cambiendo la línea correspondiente u otra.

Carlosmapego8 reviewed (changes requested):
Te dejo un par de comentarios, modificando esto tienes mi aprobado

Carlosmapego8 reviewed (approved):
Veo correcto lo que propones @FerniCuesta

FerniCuesta commented:
Buenas @JJ,

Le solicito una revisión del objetivo.

JJ commented on internal/auxiliar.go:
```diff
@@ -0,0 +1,11 @@
+	Contrato       Contrato
+	Turnos         map[time.Time]Turno
+	TurnoPreferido TipoTurno
+	DiasBaja       int
```
¿Es un objeto valor esto o es una entidad?

JJ commented on internal/contrato.go:
```diff
@@ -0,0 +1,32 @@
+type TipoContrato string
+
+const (
+	Rotatorio       TipoContrato = "Rotatorio"
```
¿De qué forma esto es una solución mínima a "los turnos deben respetar las horas máximas que pueden trabajar"? ¿Habéis entendido el concepto de *mínimamente* viable?

JJ reviewed (changes requested):
Ni la modelización ni el propio lenguaje del modelo parece que se ajuste a los issues ni a las historias de usuario. @Carlosmapego8 , este va a ser código con el que tienes que trabajar más adelante. Si incluyes cosas que no vas a necesitar en el objetivo 4, tendrás que refactorizarlas o borrarlas. ¿Quieres tener ese trabajo que hacer más adelante?
Por otro lado, os reitero la necesidad de seguir el procedimiento correcto a la hora de desarrollar. No se pone un issue en un commit "porque lo diga el profe", se hace porque es la única forma de garantizar que se cumplan los deseos del cliente expresados en historias de usuario.
Finalmente, el que se cierren los comentarios con un commit no es capricho, sino simplemente la única forma de no hacer perder el tiempo a quien revisa viendo cómo se han resuelto los comentarios qeu aparecen cerrados. Muchos comentarios cerrados sin haber cambiado la línea sobre la que se hacen es indicativo de un mal proceso de desarrollo (y en algunos casos también de revisión, al haber hecho comentarios donde no se debía)

JLombar commented on internal/contrato.go:
```diff
@@ -0,0 +1,32 @@
+// Constructor de Contrato
+func NewContrato(tipo TipoContrato, horas int, diasLibres DiasLibres) Contrato {
+
+	if horas < 0 {
```
Al igual que ha indicado JJ, sería interesante que no solo pongas una condición a las horas mínimas del contrato, si no que también añadas las horas máximas

JLombar commented on internal/contrato.go:
```diff
@@ -0,0 +1,32 @@
+package gestorturnos
+
+// Contrato
```
Si no estoy equivocado, en ningún momento se te pide que añadas comentarios para explicar las clases. Si lo vas a hacer al menos haz que siga el formato estándar de Go, pero como he dicho no deberías añadirlos si no se indica en nigún milestone ni nada por el estilo, sólo añaden deuda técnica y trabajo a futuro

JLombar commented on internal/gestorturnos.go:
```diff
@@ -0,0 +1,20 @@
+package gestorturnos
+
+import (
```
Si sólo vas a hacer un import podrías ponerlo en una línea

JLombar reviewed (commented):
Los cambios que te indico, son detalles, pero todo importa en programación. Dale otra vueltecilla y suerte
